app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,365 @@
+
+import static org.kiwix.kiwixmobile.utils.StyleUtils.dialogStyle;
+import static org.kiwix.kiwixmobile.webserver.WebServerHelper.isStarted;
+import static org.kiwix.kiwixmobile.webserver.WebServerHelper.stopAndroidWebServer;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.checkHotspotState;
+
+public class ZimHostActivity extends AppCompatActivity implements
+    ServerStateListener {
+
+  Button startServerButton;
+  TextView serverTextView;
+
+  public static final String ACTION_TURN_ON_AFTER_O = "Turn_on_hotspot_after_oreo";
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  private final String IP_STATE_KEY = "ip_state_key";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  private Intent serviceIntent;
+  private Task<LocationSettingsResponse> task;
+  boolean flag = false;
+  String ip;
+  String TAG = ZimHostActivity.this.getClass().getSimpleName();
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    setUpToolbar();
+
+    startServerButton = (Button) findViewById(R.id.startServerButton);
+    serverTextView = (TextView) findViewById(R.id.server_textView);
+
+    if (savedInstanceState != null) {
+      serverTextView.setText(
+          getString(R.string.server_started_message) + " " + savedInstanceState.getString(
+              IP_STATE_KEY));
+      startServerButton.setText(getString(R.string.stop_server_label));
+    }
+
+    FragmentManager fragmentManager = getSupportFragmentManager();
+    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
+    ZimFileSelectFragment fragment = new ZimFileSelectFragment();
+    fragmentTransaction.add(R.id.frameLayoutServer, fragment);
+    fragmentTransaction.commit();
+
+    serviceIntent = new Intent(this, HotspotService.class);
+    Context context = this;
+
+    startServerButton.setOnClickListener(new View.OnClickListener() {
+      @Override public void onClick(View v) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          toggleHotspot();
+        } else {
+          //TO DO: show Dialog() + within that add check mobile Data check later.
+          //if (isMobileDataEnabled(context)) {
+          //  mobileDataDialog();
+          //} else {
+          if (flag) {
+            serverStopped();
+            isStarted = false;
+          } else {
+            startHotspotDialog();
+          }
+          //}
+        }
+      }
+    });
+  }
+
+  @RequiresApi(api = Build.VERSION_CODES.O)
+  void toggleHotspot() {
+    //Check if location permissions are granted
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+        == PackageManager.PERMISSION_GRANTED) {
+      if (checkHotspotState(this)) //If hotspot is already enabled, turn it off
```

Comment:
`checkHotspotState` doesn't tell me what this method does hence the comment to explain it.
`isHotspotEnabled` gives more information.
The static method also worries me because it implies you have static state in your service

Reply:
Fixed. 7bbc07d3bff5295e128f046b386ec636bcc0834b

Reply:
5b7657902244d3264cbd9fe4436d38224a5a107f